Tooltip: add `placement` prop to replace deprecated `position`

brookewp opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

Part of #44401

## What?
<!-- In a few words, what is the PR actually doing? -->

Similar to [changes in Popover](https://github.com/WordPress/gutenberg/issues/44339), the use of `floating-ui` in Tooltip brings a new prop for positioning, `placement`. 

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Currently, Tooltip uses the `position` prop, which we will be replacing with `placement`. This starts the deprecation process so we can slowly introduce `placement` to avoid breaking changes. 

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

By exposing `placement`, updating the logic to favor `placement` over `position`, adding deprecation warnings, updating types and the README. As well as updating internal uses of `position` to `placement`. 

## Testing Instructions

Ensure unit tests are passing and don't show unexpected `console.warn` errors 

### Smoke tests 

`npm run storybook:dev`

#### Tooltip

- Select a `position` for tooltip and check that the error appears in the console 
- Choose a different option for `placement` to ensure it takes precedence over `position`. 

#### Button

- Enable `showTooltip` and add a `label`. 
- Ensure the default is `bottom`. 
- Try changing `tooltipPosition` to other options to ensure the tooltip displays where expected. 

#### Others

- Smoke test other components where the `position` was changed to `placement`. **Ensure that results match trunk.** I highlight this because some positions appear incorrect, but the behavior is the same. i.e. `ToggleGroupControl`

What `position: top` looks like in trunk: 

<img width="420" alt="Screenshot 2023-09-13 at 1 44 22 PM" src="https://github.com/WordPress/gutenberg/assets/35543432/67298b7d-28ac-4c2f-9149-e34f87e0aa51">

And if we experiment and change to `bottom right`: 
<img width="451" alt="Screenshot 2023-09-13 at 1 44 09 PM" src="https://github.com/WordPress/gutenberg/assets/35543432/e29cc6dc-750c-4dba-a4b7-60177a8388b0">

Something strange is happening here, so I'll add a to-do item for it. 

- i.e. `BoxControl`* (unlinked side units) and `DateTimePicker` (UTC text) can be tested in storybook. 

<img width="312" alt="Screenshot 2023-09-13 at 1 28 54 PM" src="https://github.com/WordPress/gutenberg/assets/35543432/aa38e2e3-2ede-4975-9d92-1921a11185c7">

<hr/>

***Side note**: Will also add a to-do follow up for `BoxControl`:
First, in the screenshot above, the tooltip overlaps, which doesn't seem ideal. But also, a custom implementation of Tooltip has been added as `BaseTooltip` which I believe is worth looking into.

brookewp committed:
Tooltip: add `placement` prop to replace `position`

tyxla commented on packages/components/src/tooltip/index.tsx:
```diff
@@ -44,8 +46,29 @@ function Tooltip( props: TooltipProps ) {
+	}
+	computedPlacement = computedPlacement || DEFAULT_PLACEMENT;
+
+	if ( position !== undefined ) {
+		deprecated( '`position` prop in wp.components.tooltip', {
+			since: '6.4',
+			alternative: '`placement` prop',
+		} );
+	}
```
We have this `if ( position !== undefined )` condition above already, so perhaps we can just move the `deprecated()` call there, instead of duplicating the check?

brookewp replied:
I borrowed [the logic from the same changes made to URLPopover](https://github.com/WordPress/gutenberg/pull/44391/files). Do you think it's worth following up there as well with your suggestions?

tyxla replied:
Yes. Feel free to follow up in another PR, although it's not a huge deal, just a minor code style preference.

tyxla commented on packages/components/src/tooltip/index.tsx:
```diff
@@ -44,8 +46,29 @@ function Tooltip( props: TooltipProps ) {
+	} else if ( position !== undefined ) {
+		computedPlacement = positionToPlacement( position );
+	}
+	computedPlacement = computedPlacement || DEFAULT_PLACEMENT;
```
I think we could just inline the default placement, no need to have a constant when it's only used once and in just a few lines below in the same function.

tyxla commented on packages/components/src/tooltip/README.md:
```diff
@@ -35,12 +35,21 @@ The amount of time in milliseconds to wait before showing the tooltip.
 
+#### `placement`: `'top' | 'top-start' | 'top-end' | 'right' | 'right-start' | 'right-end' | 'bottom' | 'bottom-start' | 'bottom-end' | 'left' | 'left-start' | 'left-end'`
+
+Used to specify the tooltip's position with respect to its anchor.
```
We could add a reference to the placement docs:

https://floating-ui.com/docs/tutorial#placements

brookewp replied:
I grabbed this from Popover's README for consistency: https://github.com/WordPress/gutenberg/blob/ab97d28a55b028139b66b8199d795445d830cad8/packages/components/src/popover/README.md?plain=1#L199-L201

So maybe we'd want to add the link there, too? And after mentioning URLPopover above, I looked at the README on `placement`:

https://github.com/WordPress/gutenberg/blob/ab97d28a55b028139b66b8199d795445d830cad8/packages/block-editor/src/components/url-popover/README.md?plain=1#L97-L103

So I wonder if we should follow up with these and other components moving towards `placement`.

tyxla replied:
Sounds like a good documentation improvement 👍 If these are coming from an external library, then IMHO it makes a lot of sense to link to the original library's documentation.

ciampo replied:
The counter-argument to that, is that we're exposing our internal implementation in our docs. We usually avoid doing this as it gives us more flexibility in changing underlying implementation (like when `Popover` was refactored from `popper` to `floating-ui` last year)

tyxla commented on packages/components/src/tooltip/index.tsx:
```diff
@@ -44,8 +46,29 @@ function Tooltip( props: TooltipProps ) {
+	computedPlacement = computedPlacement || DEFAULT_PLACEMENT;
+
+	if ( position !== undefined ) {
+		deprecated( '`position` prop in wp.components.tooltip', {
```
We'll either need to update some tests to expect the `deprecated` call, or to stop using `position` anywhere, otherwise a bunch of tests will fail.

brookewp replied:
Sounds good! I've started to replace our internal use of `position`, but I've run into a few things along the way ([example](https://github.com/WordPress/gutenberg/pull/54263)). In the meantime, I've set this as a draft while I look into them.

tyxla replied:
Sure, just wanted to point it out. Thanks!

brookewp replied:
Good thing you did, as we're going to update the tests in this PR and then replace internal use cases in the next PR. 😄

tyxla reviewed (commented):
Nice 👍 I'm just leaving some drive-by comments.

brookewp committed:
Refine code based on feedback

brookewp committed:
Add details on floating-ui to readme

brookewp committed:
Update tests to expect warning

brookewp commented:
@tyxla @ciampo - I've updated some of the failing tests to include an expect warning assertion. There are two more failing tests which still need to be updated but:
1. I wanted to ensure [the first few tests I've updated so far](https://github.com/WordPress/gutenberg/pull/54264/commits/bd8e7fc825871b0f0cec8f41332b4213c4f4932b) is the expected way to handle adding the deprecation warning.
2. The failing tests are both `test.each` cases. The first set of conditions expects a warning, and the other doesn't. For `packages/components/src/font-size-picker/test/index.tsx`, we could easily separate the `test.each` into single tests. 

```diff
diff --git a/packages/components/src/font-size-picker/test/index.tsx b/packages/components/src/font-size-picker/test/index.tsx
index 0d7816399b..b83a4a11be 100644
--- a/packages/components/src/font-size-picker/test/index.tsx
+++ b/packages/components/src/font-size-picker/test/index.tsx
@@ -36,32 +36,48 @@ describe( 'FontSizePicker', () => {
 		}
 	);
 
-	test.each( [
-		// Use units when first font size uses units.
-		{ firstFontSize: '12px', expectedValue: '80px' },
-		// Don't use units when first font size does not use units.
-		{ firstFontSize: 12, expectedValue: 80 },
-	] )(
-		'should call onChange( $expectedValue ) after user types 80 when first font size is $firstFontSize',
-		async ( { firstFontSize, expectedValue } ) => {
-			const user = userEvent.setup();
-			const onChange = jest.fn();
-			render(
-				<FontSizePicker
-					__nextHasNoMarginBottom
-					fontSizes={ [ { slug: 'slug', size: firstFontSize } ] }
-					onChange={ onChange }
-				/>
-			);
-			await user.click(
-				screen.getByRole( 'button', { name: 'Set custom size' } )
-			);
-			const input = screen.getByLabelText( 'Custom' );
-			await user.type( input, '80' );
-			expect( onChange ).toHaveBeenCalledTimes( 2 ); // Once per keystroke.
-			expect( onChange ).toHaveBeenCalledWith( expectedValue );
-		}
-	);
+	test( 'should call onChange( 80px ) after user types 80 when first font size is 12px', async () => {
+		const user = userEvent.setup();
+		const onChange = jest.fn();
+		render(
+			<FontSizePicker
+				__nextHasNoMarginBottom
+				fontSizes={ [ { slug: 'slug', size: '12px' } ] }
+				onChange={ onChange }
+			/>
+		);
+		await user.click(
+			screen.getByRole( 'button', { name: 'Set custom size' } )
+		);
+		const input = screen.getByLabelText( 'Custom' );
+
+		await user.type( input, '80' );
+
+		expect( console ).toHaveWarned();
+		expect( onChange ).toHaveBeenCalledTimes( 2 ); // Once per keystroke.
+		expect( onChange ).toHaveBeenCalledWith( '80px' );
+	} );
+
+	test( 'should call onChange( 80 ) after user types 80 when first font size is 12', async () => {
+		const user = userEvent.setup();
+		const onChange = jest.fn();
+		render(
+			<FontSizePicker
+				__nextHasNoMarginBottom
+				fontSizes={ [ { slug: 'slug', size: 12 } ] }
+				onChange={ onChange }
+			/>
+		);
+		await user.click(
+			screen.getByRole( 'button', { name: 'Set custom size' } )
+		);
+		const input = screen.getByLabelText( 'Custom' );
+
+		await user.type( input, '80' );
+
+		expect( onChange ).toHaveBeenCalledTimes( 2 ); // Once per keystroke.
+		expect( onChange ).toHaveBeenCalledWith( 80 );
+	} );
 
 	describe( 'with > 5 homogeneous font sizes', () => {
 		const fontSizes = [
```

But in `packages/components/src/toggle-group-control/test/index.tsx`, it gets a bit messier. The easiest solution would be to add a conditional statement, but I saw it's not good practice to put expects in conditionals. I have seen a handful of tests in GB with an ignore ` // eslint-disable-next-line jest/no-conditional-expect`, but I wanted to check to see if there were any other approaches I've missed. Although it would be as simple as adding the following (with a different condition for FontSizePicker` of course): 

```diff
diff --git a/packages/components/src/toggle-group-control/test/index.tsx b/packages/components/src/toggle-group-control/test/index.tsx
index 78be0e89fc..d685833628 100644
--- a/packages/components/src/toggle-group-control/test/index.tsx
+++ b/packages/components/src/toggle-group-control/test/index.tsx
@@ -98,6 +98,11 @@ describe.each( [
 				</Component>
 			);
 
+			if ( mode === 'uncontrolled' ) {
+				// eslint-disable-next-line jest/no-conditional-expect
+				expect( console ).toHaveWarned();
+			}
+
 			expect( container ).toMatchSnapshot();
 		} );
 	} );

```

ciampo commented:
This is what i'd do:

- I would update, in this PR, all consumers of `Tooltip` to use the `placement` prop instead of `position`:
  - when possible, we simply update the hardcoded values (for example, [`position="top center"`](https://github.com/WordPress/gutenberg/blob/10859af537871d0010f7e904705a1672aad9cc78/packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx#L41) becomes `placement="top"`
  - if, instead, the consumer is forwarding a prop (like `Button`, that [forwards the value of `tooltipPosition`](https://github.com/WordPress/gutenberg/blob/a3af2c7d0aee82a0cb8f74beae95730849b5afc6/packages/components/src/button/index.tsx#L251)), then we can use the `positionToPlacement` utility function, and do something like `placement={positionToPlacement(tooltipPosition)}`
- One we're done refactoring the consumers of `Tooltip`, the only tests that would need the `expect( console ).toHaveWarned();` assertion should be the `Tooltip` tests (which should also avoid the issues flagged in the [comment above](https://github.com/WordPress/gutenberg/pull/54264#issuecomment-1716469422))
- As a follow-up to this PR, we can update those unit tests to also use the `placement` prop, and thus remove the `expect( console ).toHaveWarned();` assertions

tyxla reviewed (commented):
Although I don't have strong feelings about it (we could alternatively follow up in additional PRs to replace all `position` with `placement` usages), I agree with the plan @ciampo proposed above, because that will help remove some of the warning inconsistencies we're seeing in the tests.

brookewp committed:
Revert "Update tests to expect warning"

This reverts commit 1f6fa73b93320c64671ad739a77116253a0b33fe.

brookewp committed:
Replace internal Tooltip usage of `position` with `placement`

brookewp commented:
Thanks for the comments! I was hoping to replace internal usage in a second PR, but it seems cleaner to do it in this PR. That way we avoid adding warnings and rewriting tests temporarily to then be reverted in the follow-up PR.

brookewp commented:
@ciampo I noticed in [similar instances ](https://github.com/WordPress/gutenberg/pull/44391)where we added `placement`, we didn't add a CHANGELOG entry. I would have thought it would be under internal, but I'll wait on you to confirm which way to go.

ciampo commented:
> @ciampo I noticed in [similar instances ](https://github.com/WordPress/gutenberg/pull/44391)where we added `placement`, we didn't add a CHANGELOG entry.

The specific example of URLPopover didn't require a CHANGELOG entry because that component is not part of the `@wordpress/components` package

>  I would have thought it would be under internal, but I'll wait on you to confirm which way to go.

Since we're adding a new public API to the `Tooltip` component, I'd argue that it not an internal change — I'd probably list it under a `### New Features` section.

ciampo commented on packages/components/src/button/index.tsx:
```diff
@@ -225,6 +226,13 @@ export function UnforwardedButton(
 			</button>
 		);
 
+	// use floating-ui's `placement` instead of legacy `position`
```
```suggestion
	// Convert legacy `position` values to be used with the new `placement` prop
```

ciampo commented on packages/components/src/tooltip/types.ts:
```diff
@@ -24,9 +29,19 @@ export type TooltipProps = {
 	 * `"bottom"` y axis, and `"left"`, `"center"`, `"right"` x axis.
+	 * Where the tooltip should be positioned relative to its parent.
 	 *
 	 * @default bottom
```
Let's add a `@deprecated` tag to the `position` prop

ciampo reviewed (commented):
Left a couple of minor comments, but changes LGTM.

@chad1008 , would you be able to take a final look at this PR, including testing the updated usages of `Tooltip` across the block editor? Feel free to approve if things look good on your end

chad1008 commented on packages/components/src/tooltip/index.tsx:
```diff
@@ -45,8 +47,24 @@ function Tooltip( props: TooltipProps ) {
+	// Compute tooltip's placement:
+	// - give priority to `placement` prop, if defined
+	// - otherwise, compute it from the legacy `position` prop (if defined)
+	// - finally, fallback to the DEFAULT_PLACEMENT.
```
Now that we're inlining the default value, we can update this comment to say `bottom` instead of `DEFAULT_PLACEMENT`

brookewp replied:
Good catch!

chad1008 reviewed (approved):
This looks good and tests well to me! I left one small comment of my own, and it looks like @ciampo may have added a couple thoughts on wording/exposing internal implementation details, but that's all code comment stuff that shouldn't be any kind of blocker 😄 🚀 🚢

brookewp commented:
>Since we're adding a new public API to the Tooltip component, I'd argue that it not an internal change — I'd probably list it under a ### New Features section.

@ciampo do you think it would make sense for `hideOnClick` to be under ### New Features as well? I could move it when updating the changelog here

brookewp committed:
Documentation updates based on feedback

tyxla commented:
Noting that something went off with the latest [merge from trunk](96d1cb148645475c3e4f72b6e495af51920c4420) - there seem to be some extra changes that we likely don't want as part of the PR.

ciampo commented:
> do you think it would make sense for `hideOnClick` to be under ### New Features as well? I could move it when updating the changelog here

Yup, sounds good to me.

> Noting that something went off with the latest [merge from trunk](96d1cb148645475c3e4f72b6e495af51920c4420) - there seem to be some extra changes that we likely don't want as part of the PR.

Good catch. Let's fix that before merging.

brookewp commented:
>Noting that something went off with the latest [merge from trunk](https://github.com/WordPress/gutenberg/pull/96d1cb148645475c3e4f72b6e495af51920c4420) - there seem to be some extra changes that we likely don't want as part of the PR.

Great find! Thank you for looking out. 😄 I am trying to be friends with `git merge` over `git rebase`, but... 😕 

I'll get that resolved along with the conflict and update the changelog.

brookewp committed:
Move other tooltip change to new features in changelog

ciampo commented on packages/components/src/tooltip/types.ts:
```diff
@@ -24,10 +29,21 @@ export type TooltipProps = {
+	 *
 	 * Specify y- and x-axis as a space-separated string. Supports `"top"`,
 	 * `"bottom"` y axis, and `"left"`, `"center"`, `"right"` x axis.
+	 * Where the tooltip should be positioned relative to its parent.
```
Probably this line is extra?

ciampo reviewed (approved):
Let's wrap this up and merge 🚀

tyxla reviewed (approved):
Looking great from my perspective too! 🚀

brookewp committed:
Remove extra comment from types
